crates/sui-sdk/examples/utils.rs

```diff
@@ -0,0 +1,305 @@
+        .collect::<Vec<_>>();
+    let recipient = addresses
+        .get(0)
+        .expect("Cannot get the recipient address needed for writing operations. Aborting");
+
+    Ok((client, active_address, *recipient))
+}
+
+/// Return a sui client to interact with the APIs and an active address from the local wallet.
+///
+/// This function sets up a wallet in case there is no wallet locally,
+/// and ensures that the active address of the wallet has SUI on it.
+/// If there is no SUI owned by the active address, then it will request
+/// SUI from the faucet.
+pub async fn setup_for_read() -> Result<(SuiClient, SuiAddress), anyhow::Error> {
+    let client = SuiClientBuilder::default().build_testnet().await?;
+    println!("Sui testnet version is: {}", client.api_version());
+    let mut wallet = retrieve_wallet().await?;
+    assert!(wallet.get_addresses().len() >= 2);
+    let active_address = wallet.active_address()?;
+
+    println!("Local wallet active address is: {active_address}");
+
+    // make sure we have some SUI (5_000_000 MIST) on this address
+    let coin = fetch_coin(&client, &active_address).await?;
+
+    if coin.is_none() {
+        request_tokens_from_faucet(active_address).await?;
+    }
+
+    Ok((client, active_address))
+}
+
+/// Request tokens from the Faucet for the given address
+pub async fn request_tokens_from_faucet(address: SuiAddress) -> Result<(), anyhow::Error> {
+    let address_str = address.to_string();
+    let json_body = json![{
+        "FixedAmountRequest": {
+            "recipient": &address_str
+        }
+    }];
+
+    // make the request to the faucet JSON RPC API for coin
+    let client = Client::new();
+    let resp = client
+        .post(SUI_FAUCET)
+        .header("Content-Type", "application/json")
+        .json(&json_body)
+        .send()
+        .await?;
+    println!(
+        "Faucet request for address {address_str} has status: {}",
+        resp.status()
+    );
+    println!("Waiting for the faucet to complete the gas request ...");
+    let faucet_resp: FaucetResponse = resp.json().await?;
+    let task_id = if faucet_resp.error.is_none() {
+        faucet_resp.task
+    } else {
+        return Err(anyhow!(
+            "Faucet request was unsuccessful. Error is {:?}",
+            faucet_resp.error
+        ));
+    };
+
+    println!("Faucet request task id: {task_id}");
+
+    let json_body = json![{
+        "GetBatchSendStatusRequest": {
+            "task_id": &task_id
+        }
+    }];
+
+    loop {
+        let resp = client
+            .get("https://faucet.testnet.sui.io/v1/status")
+            .header("Content-Type", "application/json")
+            .json(&json_body)
+            .send()
+            .await?;
+        let text = resp.text().await?;
+        if text.contains("SUCCEEDED") {
+            std::thread::sleep(Duration::from_secs(5));
+            break;
+        } else {
+            std::thread::sleep(Duration::from_secs(5));
+        }
```

Comment:
Instead of using `std::thread::sleep` use `tokio::sleep` -- this is an `async` function, so the tokio runtime is cooperatively scheduling tasks across multiple machine threads.  If you `thread::sleep`, you lock-up the thread for all tasks.  If you `tokio::sleep`, the tokio runtime can park the task for the requisite time.

```suggestion
            tokio::time::sleep(Duration::from_secs(5)).await;
            break;
        } else {
            tokio::time::sleep(Duration::from_secs(5)).await;
        }
```

Reply:
Also, 5 seconds seems like quite a long poll time, try 1s?